### PR TITLE
Implement mania composer editor screen

### PR DESCRIPTION
**Pull request by Rekkonnect**
This implements the composer editor screen for osu!mania beatmaps. The following changes have been made overall:

- Add osu!mania playfield without showing judgements
- Add snap lines according to the current beat snap divisor (using osu!stable's colours)
- Test case

TODO:
- [ ] Disable key input (only when not live mapping)
- [ ] Let hit objects appear above the snap lines
- [ ] Maintain a fixed scrolling speed dependent on the timeline's scrolling speed
- [x] Ensure no duplicate snap lines are created
- [ ] Add mask on top of the selected notes (split to [#2602](https://github.com/ppy/osu/pull/2602))
- [ ] Select notes
- [ ] Move notes
- [ ] Adjust hold notes' end position
- [ ] Create notes
- [ ] Delete notes
- [ ] Copy-paste notes

### CODE REVIEW

**Review by smoogipoo — CHANGES_REQUESTED** [review]
Initial.

File: osu.Game.Rulesets.Mania/UI/ManiaStage.cs [hunk 1]
```diff
@@ -169,8 +174,21 @@ public override void Add(DrawableHitObject h)
...
 
+        public void Add(EditSnapLine editSnapLine) => base.Add(new DrawableEditSnapLine(editSnapLine));
+        public void Remove(DrawableEditSnapLine editSnapLine) => base.Remove(editSnapLine);
+        public void ClearEditSnapLines()
+        {
+            var o = HitObjects.Objects;
+            for (int i = o.Count() - 1; i >= 0; i--)
+                if (o.ElementAt(i) is DrawableEditSnapLine e)
+                    Remove(e);
+        }
```
[the remark this answers is not in the document]
Comment (AlFasGD) [reply osu.Game.Rulesets.Mania/UI/ManiaStage.cs]: I don't really get what you're asking, could you do these changes yourself?

**Comment by smoogipoo** [comment]
Please split out the mask stuff into a separate PR.

**Comment by smoogipoo** [comment]
Closing for now to reduce clutter - this PR isn't going to go in in its current form in any case as each addition should be split out into its own PR. A significant portion of it is in https://github.com/ppy/osu/pull/2602 .

**Pull request closed without merging**